data/tutorials/language/0lg_08_basic_datatypes.md

```diff
@@ -0,0 +1,686 @@
+
+
+#### Recursive Variants
+
+A variant definition referring to itself is recursive. A constructor may wrap data from the type being defined.
+
+This is the case of the following definition, which can be used to store JSON values.
+```ocaml
+# type json =
+  | Null
+  | Bool of bool
+  | Int of int
+  | Float of float
+  | String of string
+  | Array of json list
+  | Object of (string * json) list;;
+type json =
+    Null
+  | Bool of bool
+  | Int of int
+  | Float of float
+  | String of string
+  | Array of json list
+  | Object of (string * json) list
+```
+
+Both constructors `Array` and `Object` contain values of type `json`.
+
+Functions defined using pattern matching on recursive variants are often recursive too. This function checks if a name is present in a whole JSON tree:
+```ocaml
+# let rec has_field name = function
+  | Array u ->
+      List.fold_left (fun b obj -> b || has_field name obj) false u
+  | Object u ->
+      List.fold_left (fun b (key, obj) -> b || key = name || has_field name obj) false u
+  | _ -> false;;
+val has_field : string -> json -> bool = <fun>
+```
+
+Here, the last pattern uses the symbol `_`, which catches everything. It allows returning `false` on all data which is neither `Array` nor `Object`.
+
+### Polymorphic Data Types
+
+#### Revisiting Predefined Types
+
+The predefined type `option` is defined as a variant type with two constructors: `Some` and `None`. It can contain values of any type, such as `Some 42` or `Some "hola"`. The variant `option` is polymorphic. Here is how it is defined in the standard library:
+
+```ocaml
+# #show option;;
+type 'a option = None | Some of 'a
+```
+
+The predefined type `list` is also a polymorphic variant with two constructors. Here is how it is defined in the standard library:
+```ocaml
+# #show list;;
+type 'a list = [] | (::) of 'a * 'a list
+```
+
+The only bit of magic here is turning constructors into symbols. This is left unexplained in this tutorial. The types `bool` and `unit` also are regular variants, with the same magic:
+```ocaml
+# #show unit;;
+type unit = ()
+
+# #show bool;;
+type bool = false | true
+```
+
+Implicitly, product types also behave as variant types. For instance, pairs can be seen as inhabitants of this type:
+```ocaml
+# type ('a, 'b) pair = Pair of 'a * 'b;;
+type ('a, 'b) pair = Pair of 'a * 'b
+```
+
+Where `(int, bool) pair` would be written `int * bool` and `Pair (42, true)` would be written `(42, true)`. From the developer's perspective, everything happens as if such a type would be declared for every possible product shape. This is what allows pattern matching on products.
+
+Even integers and floats can be seen as enumerated-like variant types, with many constructors and funky syntactic sugar. This is what allows pattern matching on those types.
+
+In the end, the only type construction that does not reduce to a variant is the function arrow type. No pattern matching on functions.
```

Comment:
Do you mean "There is no pattern matching on functions" for the second sentence?

Reply:
> It possible to perform pattern matching on values of any type, except functions.